dragnet/compat.py

```diff
@@ -8,20 +8,259 @@
     string_ = (str, unicode)
+    from itertools import izip as zip_
     import cPickle as pickle
     from StringIO import StringIO as bytes_io
 else:
     range_ = range
     bytes_ = bytes
     unicode_ = str
     string_ = (bytes, str)
+    zip_ = zip
     import pickle
     from io import BytesIO as bytes_io
 
+
+def str_cast(maybe_bytes, encoding='utf-8'):
+    """
+    Converts any bytes-like input to a string-like output, with respect to python version
+
+    Parameters
+    ----------
+    maybe_bytes : if this is a bytes-like object, it will be converted to a utf-8 string
+    encoding  : str, default='utf-8'
+        encoding to be used when decoding bytes
+    """
+    if isinstance(maybe_bytes, bytes_):
+        return maybe_bytes.decode(encoding)
+    else:
+        return maybe_bytes
+
+
+def bytes_cast(maybe_str, encoding='utf-8'):
+    """
+    Converts any string-like input to a bytes-like output, with respect to python version
+
+    Parameters
+    ----------
+    maybe_str : if this is a string-like object, it will be converted to bytes(assumes utf-8 encoding)
+    encoding  : str, default='utf-8'
+        encoding to be used when encoding string
+    """
+    if isinstance(maybe_str, unicode_):
+        return maybe_str.encode(encoding)
+    else:
+        return maybe_str
+
+
+def str_list_cast(list_, **kwargs):
+    """
+    Converts any bytes-like items in input list to string-like values, with respect to python version
+
+    Parameters
+    ----------
+    list_ : list
+        any bytes-like objects contained in the list will be converted to strings (assumed utf-8 encoding)
+    kwargs:
+        encoding: str, default: 'utf-8'
+            encoding to be used when decoding bytes
+    """
+    return [str_cast(elem, **kwargs) for elem in list_]
+
+
+def bytes_list_cast(list_, **kwargs):
+    """
+    Converts any string-like items in input list to bytes-like values, with respect to python version
+
+    Parameters
+    ----------
+    list_ : list
+        any string-like objects contained in the list will be converted to bytes (assumed utf-8 encoding)
+    kwargs:
+        encoding: str, default: 'utf-8'
+            encoding to be used when encoding string
+    """
+    return [bytes_cast(elem, **kwargs) for elem in list_]
+
+
+def str_dict_cast(dict_, include_keys=True, include_vals=True, **kwargs):
+    """
+    Converts any bytes-like items in input dict to string-like values, with respect to python version
+
+    Parameters
+    ----------
+    dict_ : dict
+        any bytes-like objects contained in the dict will be converted to a string (assumed utf-8 encoding)
+    include_keys : bool, default=True
+        if True, cast keys to a string, else ignore
+    include_values : bool, default=True
+        if True, cast values to a string, else ignore
+    kwargs:
+        encoding: str, default: 'utf-8'
+            encoding to be used when decoding bytes
+    """
+    old_keys = dict_.keys()
```

Comment:
Nit: I don't think there's a need to assign these two old_ variables, since you can just do, e.g. `new_keys = str_list_cast(dict_.keys(), **kwargs) if include_keys else dict_.keys()`. Am assuming there's no real performance difference, I just found it more confusing to read the code.